lib/workflow/stores/mongo.js

```diff
@@ -643,18 +643,34 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
             graphId: data.graphId,
             reachable: true
         };
-        query['dependencies.' + data.taskId] = {
-            $in: [data.state, Constants.Task.States.Finished]
-        };
         var update = {
             $unset: {}
         };
-        update.$unset['dependencies.' + data.taskId] = '';
         var options = {
             multi: true
         };
+        
+        query['dependencies.' + data.taskId] = {
+            $in: [data.state, Constants.Task.States.Finished]
+        };
 
-        return waterline.taskdependencies.updateMongo(query, update, options);
+        return waterline.taskdependencies.find(query).then(function(res) {
+           if (_.isEmpty(res)) {
+                // no task found under dependencies try search key anyOf
+                delete query['dependencies.' + data.taskId];
+                query['dependencies.anyOf.' + data.taskId] = {
+                    $in: [data.state, Constants.Task.States.Finished]
+                };
+
+                update.$unset['dependencies.anyOf'] = '';
```

Comment:
Sorry, it's my fault, your code is correct, since any condition is meet, then the whole `anyOf` condition is meet, so to delete the whole `anyOf` is reasonable.